File: app/public/js/synths/tb303.js
/**
 * TB-303 Bass Synthesizer
 * Classic acid bassline synthesis with resonant filter
 */

class TB303 {
    constructor(audioContext) {
        this.context = audioContext;
        this.output = this.context.createGain();
        this.output.gain.value = 0.8;

        // TB-303 parameters
        this.params = {
            // Oscillator
            waveform: 'sawtooth', // sawtooth or square
            tuning: 0, // cents

            // Filter
            cutoff: 800, // Hz
            resonance: 18, // 0-30
            envMod: 0.7, // envelope modulation amount (0-1)

            // Envelope
            decay: 0.4, // seconds
            accent: false,

            // Glide
            slide: false,
            slideTime: 0.1, // seconds
        };

        // Current state
        this.currentNote = null;
        this.currentFrequency = null;

        // Pattern (16 steps)
        this.pattern = [];
    }

    // Play a note
    playNote(noteData) {
        const { note, accent = false, slide = false, duration = 0.3 } = noteData;

        const frequency = this.noteToFrequency(note);
        const time = this.context.currentTime;

        // Create oscillator
        const osc = this.context.createOscillator();
        osc.type = this.params.waveform;

        // Glide/slide
        if (slide && this.currentFrequency) {
            osc.frequency.setValueAtTime(this.currentFrequency, time);
            osc.frequency.exponentialRampToValueAtTime(
                frequency,
                time + this.params.slideTime,
            );
        } else {
            osc.frequency.setValueAtTime(frequency, time);
        }

        this.currentFrequency = frequency;

        // Create filter (the iconic 303 sound)
        const filter = this.context.createBiquadFilter();
        filter.type = 'lowpass';

        // Filter cutoff with envelope
        const cutoffBase = this.params.cutoff;
        const cutoffPeak = cutoffBase + (cutoffBase * this.params.envMod * 4);
        const accentMultiplier = accent ? 1.5 : 1;

        filter.frequency.setValueAtTime(cutoffBase, time);
        filter.frequency.exponentialRampToValueAtTime(
            cutoffPeak * accentMultiplier,
            time + 0.01,
        );
        filter.frequency.exponentialRampToValueAtTime(
            cutoffBase,
            time + this.params.decay,
        );

        // Resonance
        filter.Q.value = this.params.resonance;

        // Amplitude envelope
        const amp = this.context.createGain();
        const ampPeak = accent ? 0.9 : 0.6;

        amp.gain.setValueAtTime(0, time);
        amp.gain.linearRampToValueAtTime(ampPeak, time + 0.003);
        amp.gain.exponentialRampToValueAtTime(
            0.001,
            time + duration,
        );

        // Connect: osc -> filter -> amp -> output
        osc.connect(filter);
        filter.connect(amp);
        amp.connect(this.output);

        // Start and stop
        osc.start(time);
        osc.stop(time + duration);

        this.currentNote = note;
    }

    // Set parameter
    setParam(param, value) {
        if (this.params.hasOwnProperty(param)) {
            this.params[param] = value;
            console.log(`TB-303: ${param} = ${value}`);
        }
    }

    // Note to frequency
    noteToFrequency(note) {
        const notes = {
            C: 0,
'C#': 1,
D: 2,
'D#': 3,
E: 4,
F: 5,
            'F#': 6,
G: 7,
'G#': 8,
A: 9,
'A#': 10,
B: 11,
        };

        const match = note.match(/^([A-G]#?)(\d+)$/);
        if (!match) return 440;

        const [, noteName, octave] = match;
        const noteNumber = notes[noteName];
        const octaveNumber = parseInt(octave);

        const midiNote = (octaveNumber + 1) * 12 + noteNumber;
        return 440 * 2 ** ((midiNote - 69) / 12);
    }

    // Connect to destination
    connect(destination) {
        this.output.connect(destination);
        return this;
    }

    // Disconnect
    disconnect() {
        this.output.disconnect();
        return this;
    }
}

// Make available globally
window.TB303 = TB303;

File: app/synthesis/tb303.py
"""
TB-303 Acid Bass Synthesizer - Python Backend
Generates WAV files from TB-303 patterns using NumPy/SciPy

Features:
- Classic sawtooth/square oscillator
- Resonant lowpass filter with envelope modulation
- Accent, slide, and gate controls
- 16-step sequencer rendering to audio
"""

import numpy as np
from scipy import signal
import json


class TB303:
    """TB-303 Acid Bass Synthesizer"""
    
    def __init__(self, sample_rate=44100):
        self.sample_rate = sample_rate
        
        # Default parameters
        self.params = {
            'cutoff': 800,
            'resonance': 15,
            'env_mod': 70,
            'decay': 0.3,
            'accent_level': 50,
            'waveform': 'sawtooth',
            'tuning': 0,
            'volume': 70,
            'distortion': 20,
            'delay': 0
        }
        
        # Note frequency table
        self.note_frequencies = {
            'C2': 65.41, 'C#2': 69.30, 'D2': 73.42, 'D#2': 77.78,
            'E2': 82.41, 'F2': 87.31, 'F#2': 92.50, 'G2': 98.00,
            'G#2': 103.83, 'A2': 110.00, 'A#2': 116.54, 'B2': 123.47,
            'C3': 130.81, 'C#3': 138.59, 'D3': 146.83, 'D#3': 155.56,
            'E3': 164.81, 'F3': 174.61, 'F#3': 185.00, 'G3': 196.00,
            'G#3': 207.65, 'A3': 220.00, 'A#3': 233.08, 'B3': 246.94,
            'C4': 261.63
        }
    
    def set_param(self, param, value):
        """Set synthesizer parameter"""
        if param in self.params:
            self.params[param] = float(value)
            return True
        return False
    
    def generate_oscillator(self, frequency, duration, waveform='sawtooth'):
        """Generate oscillator waveform"""
        t = np.linspace(0, duration, int(self.sample_rate * duration), False)
        
        if waveform == 'sawtooth':
            return signal.sawtooth(2 * np.pi * frequency * t)
        elif waveform == 'square':
            return signal.square(2 * np.pi * frequency * t)
        elif waveform == 'sine':
            return np.sin(2 * np.pi * frequency * t)
        else:
            return signal.sawtooth(2 * np.pi * frequency * t)
    
    def apply_filter(self, audio, cutoff_envelope, resonance):
        """Apply resonant lowpass filter with envelope"""
        output = np.zeros_like(audio)
        
        # Simple lowpass filter with envelope modulation
        for i, cutoff in enumerate(cutoff_envelope):
            # Normalize cutoff to Nyquist frequency
            normalized_cutoff = min(cutoff / (self.sample_rate / 2), 0.99)
            
            # Create filter coefficients
            b, a = signal.butter(2, normalized_cutoff, btype='low')
            
            # Apply filter to small window
            window_size = min(1024, len(audio) - i)
            if window_size > 0:
                filtered = signal.lfilter(b, a, audio[i:i+window_size])
                output[i:i+window_size] += filtered * (1.0 / len(cutoff_envelope))
        
        return output
    
    def apply_envelope(self, length, attack=0.01, decay=0.3, sustain=0.7, release=0.1):
        """Generate ADSR envelope"""
        attack_samples = int(attack * self.sample_rate)
        decay_samples = int(decay * self.sample_rate)
        release_samples = int(release * self.sample_rate)
        sustain_samples = length - attack_samples - decay_samples - release_samples
        
        if sustain_samples < 0:
            sustain_samples = 0
            attack_samples = int(length * 0.3)
            decay_samples = int(length * 0.3)
            release_samples = length - attack_samples - decay_samples
        
        # Attack
        attack_env = np.linspace(0, 1, attack_samples)
        
        # Decay
        decay_env = np.linspace(1, sustain, decay_samples)
        
        # Sustain
        sustain_env = np.ones(sustain_samples) * sustain
        
        # Release
        release_env = np.linspace(sustain, 0, release_samples)
        
        envelope = np.concatenate([attack_env, decay_env, sustain_env, release_env])
        
        # Ensure exact length
        if len(envelope) < length:
            envelope = np.pad(envelope, (0, length - len(envelope)), constant_values=0)
        elif len(envelope) > length:
            envelope = envelope[:length]
        
        return envelope
    
    def apply_distortion(self, audio, amount):
        """Apply soft clipping distortion"""
        if amount <= 0:
            return audio
        
        # Soft clipping with tanh
        gain = 1 + (amount / 100) * 4
        return np.tanh(audio * gain) / gain
    
    def synthesize_note(self, note, accent=False, slide=False, gate=False, duration=0.25):
        """Synthesize single 303 note"""
        # Get frequency
        frequency = self.note_frequencies.get(note, 130.81)
        frequency *= np.power(2, self.params['tuning'] / 1200)
        
        # Generate oscillator
        audio = self.generate_oscillator(frequency, duration, self.params['waveform'])
        
        # Calculate filter envelope
        base_cutoff = self.params['cutoff']
        env_amount = (self.params['env_mod'] / 100) * (base_cutoff * 2)
        accent_mult = 1 + (self.params['accent_level'] / 100) if accent else 1
        peak_cutoff = min(base_cutoff + (env_amount * accent_mult), 8000)
        
        # Create cutoff envelope
        decay_samples = int(self.params['decay'] * self.sample_rate)
        cutoff_envelope = np.linspace(peak_cutoff, base_cutoff, len(audio))
        
        # Apply filter (simplified - use constant cutoff for performance)
        normalized_cutoff = min(base_cutoff / (self.sample_rate / 2), 0.99)
        b, a = signal.butter(2, normalized_cutoff, btype='low')
        audio = signal.lfilter(b, a, audio)
        
        # Apply amplitude envelope
        gate_time = 0.25 if gate else 0.1
        env_length = int(min(gate_time, duration) * self.sample_rate)
        amplitude_env = self.apply_envelope(
            len(audio),
            attack=0.001,
            decay=self.params['decay'],
            sustain=0.7,
            release=0.1
        )
        
        audio *= amplitude_env
        
        # Apply accent
        if accent:
            audio *= accent_mult
        
        # Apply distortion
        if self.params['distortion'] > 0:
            audio = self.apply_distortion(audio, self.params['distortion'])
        
        # Normalize
        max_val = np.max(np.abs(audio))
        if max_val > 0:
            audio = audio / max_val * (self.params['volume'] / 100)
        
        return audio
    
    def render_pattern(self, pattern, bpm=130):
        """Render 16-step pattern to audio"""
        # Calculate step duration (16th notes)
        step_duration = (60.0 / bpm) / 4
        
        # Render each step
        audio_chunks = []
        
        for step in pattern:
            if step['active']:
                note_audio = self.synthesize_note(
                    step['note'],
                    step.get('accent', False),
                    step.get('slide', False),
                    step.get('gate', False),
                    step_duration
                )
            else:
                # Silent step
                note_audio = np.zeros(int(step_duration * self.sample_rate))
            
            audio_chunks.append(note_audio)
        
        # Concatenate all steps
        full_audio = np.concatenate(audio_chunks)
        
        # Normalize final output
        max_val = np.max(np.abs(full_audio))
        if max_val > 0:
            full_audio = full_audio / max_val * 0.9
        
        return full_audio
    
    def export_wav(self, audio, filename):
        """Export audio to WAV file"""
        try:
            from scipy.io import wavfile
            
            # Convert to 16-bit PCM
            audio_int16 = np.int16(audio * 32767)
            
            # Write WAV file
            wavfile.write(filename, self.sample_rate, audio_int16)
            return True
        except Exception as e:
            print(f"Error exporting WAV: {e}")
            return False
    
    def load_pattern_from_json(self, json_data):
        """Load pattern from JSON string or dict"""
        if isinstance(json_data, str):
            data = json.loads(json_data)
        else:
            data = json_data
        
        if 'params' in data:
            self.params.update(data['params'])
        
        return data.get('pattern', [])


# Example usage
if __name__ == '__main__':
    # Create synthesizer
    synth = TB303(sample_rate=44100)
    
    # Create classic 303 pattern
    pattern = [
        {'active': True, 'note': 'C3', 'accent': True, 'slide': False, 'gate': False},
        {'active': False, 'note': 'C3', 'accent': False, 'slide': False, 'gate': False},
        {'active': True, 'note': 'D#3', 'accent': False, 'slide': True, 'gate': False},
        {'active': False, 'note': 'C3', 'accent': False, 'slide': False, 'gate': False},
        {'active': True, 'note': 'C3', 'accent': False, 'slide': False, 'gate': False},
        {'active': False, 'note': 'C3', 'accent': False, 'slide': False, 'gate': False},
        {'active': True, 'note': 'G3', 'accent': True, 'slide': False, 'gate': True},
        {'active': False, 'note': 'C3', 'accent': False, 'slide': False, 'gate': False},
        {'active': True, 'note': 'C3', 'accent': False, 'slide': False, 'gate': False},
        {'active': False, 'note': 'C3', 'accent': False, 'slide': False, 'gate': False},
        {'active': True, 'note': 'D#3', 'accent': False, 'slide': True, 'gate': False},
        {'active': False, 'note': 'C3', 'accent': False, 'slide': False, 'gate': False},
        {'active': True, 'note': 'C3', 'accent': True, 'slide': False, 'gate': False},
        {'active': False, 'note': 'C3', 'accent': False, 'slide': False, 'gate': False},
        {'active': True, 'note': 'D3', 'accent': False, 'slide': False, 'gate': False},
        {'active': False, 'note': 'C3', 'accent': False, 'slide': False, 'gate': False}
    ]
    
    # Render pattern
    audio = synth.render_pattern(pattern, bpm=130)
    
    # Export to WAV
    synth.export_wav(audio, 'tb303_pattern.wav')
    print("Exported tb303_pattern.wav")
